src/server/src/batch.cpp

```diff
@@ -109,6 +110,202 @@ handle_edge_names_types(EdgeInsertion & in, stinger_t * S, std::string & src, st
   }
 }
 
+// Allows an EdgeInsertion to be passed to stinger_batch functions
+struct EdgeInsertionAdapter
+{
+    typedef EdgeInsertion update;
+
+    static int64_t get_type(const update &u) { return u.type(); }
+    static void set_type(update &u, int64_t v) { u.set_type(v); }
+    static int64_t get_source(const update &u) { return u.source(); }
+    static void set_source(update &u, int64_t v) { u.set_source(v); }
+    static int64_t get_dest(const update &u) { return u.destination(); }
+    static void set_dest(update &u, int64_t v) { u.set_destination(v); }
+    static int64_t get_weight(const update &u) { return u.weight(); }
+    static int64_t get_time(const update &u) { return u.time(); }
+    static int64_t get_result(const update& u) { return u.result(); }
+    static int64_t set_result(update &u, int64_t v) { u.set_result(v); }
+};
+
+template <int64_t type>
+void process_insertions(stinger_t * S, StingerBatch & batch)
+{
+    OMP("omp parallel for")
+    for (size_t i = 0; i < batch.insertions_size(); i++)
+    {
+        EdgeInsertion & in = *batch.mutable_insertions(i);
+        int64_t u = -1, v = -1;
+        std::string src, dest;
+        handle_edge_names_types<type>(in, S, src, dest, u, v);
+        if(u == -1 || v == -1) {
+            // Prevents batch update from trying to insert this edge
+            in.set_result(-1);
+        }
+    }
+
+    if (batch.make_undirected())
+    {
+        stinger_batch_incr_edge_pairs<EdgeInsertionAdapter>(
+                S, batch.mutable_insertions()->begin(), batch.mutable_insertions()->end());
+    } else {
+        stinger_batch_incr_edges<EdgeInsertionAdapter>(
+                S, batch.mutable_insertions()->begin(), batch.mutable_insertions()->end());
+    }
+
+    OMP("omp parallel for")
```

Comment:
This loop is just checking return codes and optionally doing vertexID -> name lookups. Each iteration ought to do about the same amount of work, so a static schedule will be balanced with low synchronization overhead.